File: src/modules/company/controllers/company_controller.py
from ninja import Router, Query, File
from ninja.files import UploadedFile
from src.shared.decorators.require_abilities_decorator import require_abilities
from src.modules.role.domain.enums.ability_codes_enum import AbilityCodes
from src.modules.role.domain.enums.ability_actions_enum import AbilityActions
from src.shared.domain.dto.pagination_response_dto import PaginationResponseDTO
from src.shared.types.user_request import UserRequest
from src.modules.company.domain.schemas.company_schema import (
    CompanySchema,
)
from src.modules.company.domain.dto.company_find_many_dto import (
    CompanyFindManyDTO,
)
from src.modules.company.services.company_find_many_service import (
    CompanyFindManyService,
)
from src.modules.company.services.company_find_one_service import (
    CompanyFindOneService,
)
from src.modules.company.domain.dto.company_create_dto import (
    CompanyCreateDTO,
)
from src.modules.company.services.company_create_service import (
    CompanyCreateService,
)
from src.modules.company.domain.dto.company_update_dto import (
    CompanyUpdateDTO,
)
from src.modules.company.services.company_update_service import (
    CompanyUpdateService,
)
from src.modules.company.domain.dto.company_update_status_dto import (
    CompanyUpdateStatusDTO,
)
from src.modules.company.services.company_update_status_service import (
    CompanyUpdateStatusService,
)
from src.modules.company.services.company_delete_service import (
    CompanyDeleteService,
)
from src.modules.company.services.company_update_avatar_service import (
    CompanyUpdateAvatarService,
)

from src.modules.company.services.company_delete_avatar_service import (
    CompanyDeleteAvatarService,
)

router = Router(tags=["Companies"])


@router.get("", response=PaginationResponseDTO[CompanySchema])
@require_abilities(AbilityCodes.COMPANIES, AbilityActions.READ)
def find_all(request, dto: Query[CompanyFindManyDTO]):
    user_request: UserRequest = request.user
    return CompanyFindManyService.execute(dto.dict(), user_request)


@router.get("/{id}", response=CompanySchema)
@require_abilities(AbilityCodes.COMPANIES, AbilityActions.READ)
def find_one(request, id: int):
    user_request: UserRequest = request.user
    return CompanyFindOneService.execute(id, user_request)


@router.post("", response=CompanySchema)
@require_abilities(AbilityCodes.COMPANIES, AbilityActions.CREATE)
def create(request, dto: CompanyCreateDTO):
    user_request: UserRequest = request.user
    return CompanyCreateService.execute(dto, user_request)


@router.put("/{id}", response=CompanySchema)
@require_abilities(AbilityCodes.COMPANIES, AbilityActions.UPDATE)
def update(request, id: int, dto: CompanyUpdateDTO):
    user_request: UserRequest = request.user
    return CompanyUpdateService.execute(id, dto, user_request)


@router.patch("/{id}/status", response=CompanySchema)
@require_abilities(AbilityCodes.COMPANIES, AbilityActions.UPDATE)
def update_status(request, id: int, dto: CompanyUpdateStatusDTO):
    user_request: UserRequest = request.user
    return CompanyUpdateStatusService.execute(id, dto, user_request)


@router.post("/{id}/avatar", response=CompanySchema)
@require_abilities(AbilityCodes.COMPANIES, AbilityActions.UPDATE)
def update_avatar(request, id: int, file: UploadedFile = File(...)):
    user_request: UserRequest = request.user
    return CompanyUpdateAvatarService.execute(id, file, user_request)


@router.delete("/{id}/avatar", response=CompanySchema)
@require_abilities(AbilityCodes.COMPANIES, AbilityActions.DELETE)
def delete_avatar(request, id: int):
    user_request: UserRequest = request.user
    return CompanyDeleteAvatarService.execute(id, user_request)


@router.delete("/{id}", response=CompanySchema)
@require_abilities(AbilityCodes.COMPANIES, AbilityActions.DELETE)
def delete(request, id: int):
    user_request: UserRequest = request.user
    return CompanyDeleteService.execute(id, user_request)

File: src/modules/company/controllers/company.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Patch,
  Post,
  Put,
  Query,
  UploadedFile,
  UseInterceptors,
} from '@nestjs/common';
import { FileInterceptor } from '@nestjs/platform-express';
import { ApiTags } from '@nestjs/swagger';
import { $Enums } from '@prisma/client';
import { RequireAbilities } from 'src/shared/decorators/require-abilities.decorator';
import { User } from 'src/shared/decorators/user.decorator';
import { PaginationResponseDto } from 'src/shared/domain/dto/pagination-response.dto';
import { UserRequest } from 'src/shared/types/user-request';
import { CompanyCreateDto } from '../domain/dto/company-create.dto';
import { CompanyFindManyDto } from '../domain/dto/company-find-many.dto';
import { CompanyUpdateStatusDto } from '../domain/dto/company-update-status.dto';
import { CompanyUpdateDto } from '../domain/dto/company-update.dto';
import { CompanyEntity } from '../domain/entities/company.entity';
import { CompanyCreateService } from '../services/company-create.service';
import { CompanyDeleteAvatarService } from '../services/company-delete-avatar.service';
import { CompanyDeleteService } from '../services/company-delete.service';
import { CompanyFindManyService } from '../services/company-find-many.service';
import { CompanyFindOneService } from '../services/company-find-one.service';
import { CompanyUpdateAvatarService } from '../services/company-update-avatar.service';
import { CompanyUpdateStatusService } from '../services/company-update-status.service';
import { CompanyUpdateService } from '../services/company-update.service';

@ApiTags('Companies')
@Controller('companies')
export class CompanyController {
  constructor(
    private readonly companyFindManyService: CompanyFindManyService,
    private readonly companyFindOneService: CompanyFindOneService,
    private readonly companyCreateService: CompanyCreateService,
    private readonly companyUpdateService: CompanyUpdateService,
    private readonly companyUpdateStatusService: CompanyUpdateStatusService,
    private readonly companyDeleteService: CompanyDeleteService,
    private readonly companyUpdateAvatarService: CompanyUpdateAvatarService,
    private readonly companyDeleteAvatarService: CompanyDeleteAvatarService,
  ) {}

  @Get()
  @RequireAbilities($Enums.AbilityCodes.COMPANIES, $Enums.AbilityActions.READ)
  async findAll(
    @Query() params: CompanyFindManyDto,
    @User() userRequest: UserRequest,
  ): Promise<PaginationResponseDto<CompanyEntity>> {
    return await this.companyFindManyService.execute(params, userRequest);
  }

  @Get(':id')
  @RequireAbilities($Enums.AbilityCodes.COMPANIES, $Enums.AbilityActions.READ)
  async findOne(
    @Param('id') id: number,
    @User() userRequest: UserRequest,
  ): Promise<CompanyEntity> {
    return await this.companyFindOneService.execute(id, userRequest);
  }

  @Post()
  @RequireAbilities($Enums.AbilityCodes.COMPANIES, $Enums.AbilityActions.CREATE)
  async create(
    @Body() dto: CompanyCreateDto,
    @User() userRequest: UserRequest,
  ): Promise<CompanyEntity> {
    return await this.companyCreateService.execute(dto, userRequest);
  }

  @Post(':id/avatar')
  @RequireAbilities($Enums.AbilityCodes.COMPANIES, $Enums.AbilityActions.UPDATE)
  @UseInterceptors(FileInterceptor('file'))
  async uploadAvatar(
    @Param('id') id: number,
    @UploadedFile() file: Express.Multer.File,
    @User() userRequest: UserRequest,
  ): Promise<CompanyEntity> {
    return await this.companyUpdateAvatarService.execute(id, file, userRequest);
  }

  @Delete(':id/avatar')
  @RequireAbilities($Enums.AbilityCodes.COMPANIES, $Enums.AbilityActions.DELETE)
  async deleteAvatar(
    @Param('id') id: number,
    @User() userRequest: UserRequest,
  ): Promise<CompanyEntity> {
    return await this.companyDeleteAvatarService.execute(id, userRequest);
  }

  @Put(':id')
  @RequireAbilities($Enums.AbilityCodes.COMPANIES, $Enums.AbilityActions.UPDATE)
  async update(
    @Param('id') id: number,
    @Body() dto: CompanyUpdateDto,
    @User() userRequest: UserRequest,
  ): Promise<CompanyEntity> {
    return await this.companyUpdateService.execute(id, dto, userRequest);
  }

  @Delete(':id')
  @RequireAbilities($Enums.AbilityCodes.COMPANIES, $Enums.AbilityActions.DELETE)
  async delete(
    @Param('id') id: number,
    @User() userRequest: UserRequest,
  ): Promise<CompanyEntity> {
    return await this.companyDeleteService.execute(id, userRequest);
  }

  @Patch(':id/status')
  @RequireAbilities($Enums.AbilityCodes.COMPANIES, $Enums.AbilityActions.UPDATE)
  async updateStatus(
    @Param('id') id: number,
    @Body() dto: CompanyUpdateStatusDto,
    @User() userRequest: UserRequest,
  ): Promise<CompanyEntity> {
    return await this.companyUpdateStatusService.execute(id, dto, userRequest);
  }
}
